# Data Wrangling - Tenis Case

## Tenis Case Description

### Dataframe creation - From opening the file
###### See referemces amd recomendations in: https://pandas.pydata.org/docs/user_guide/io.html

In [1]:
# Import Libraries
import pandas as pd
import numpy as np

In [2]:
# Definir path and name of the file
# path = complete if it is necesary
file_name = "Canchas 0.2.xlsx"

In [3]:
# Read the file - Load in df Dataframe
df = pd.read_excel(file_name, sheet_name="Hoja1")

### Data Observation
###### See references and recomendatons in: https://pandas.pydata.org/docs/user_guide/basics.html

In [4]:
# Check names of columns
df.columns

Index(['reserva ', 'fecha', 'hora', 'servicio', 'mza-lote', 'casa',
       'Unnamed: 6'],
      dtype='object')

In [5]:
# Check data tipes
df.dtypes

reserva                int64
fecha         datetime64[ns]
hora                  object
servicio              object
mza-lote              object
casa                  object
Unnamed: 6            object
dtype: object

Warning: Date types incorrect or not apropiate. Correction is needed. (Above)

In [6]:
# Check first 5 rows
df

reserva       fecha           hora          servicio  mza-lote    casa  \
0     8219516 2023-07-08       16:00:00  Tenis - Cancha 2  APROBADA    M-03   
1     8218607 2023-07-08       16:00:00  Tenis - Cancha 1  APROBADA    F-04   
2     8217992 2023-07-08       13:00:00  Tenis - Cancha 1  APROBADA    G-12   
3     8212103 2023-07-07       11:00:00  Tenis - Cancha 1  APROBADA    A-52   
4     8196830 2023-07-04       16:00:00  Tenis - Cancha 1  APROBADA    K-04   
..        ...        ...            ...               ...       ...     ...   
349   1712511 2020-12-04  18:00 CON LUZ  Tenis - Cancha 1  APROBADA  PRUEBA   
350   1712222 2020-12-04       17:00:00  Tenis - Cancha 1  APROBADA    E-08   
351   1687608 2020-12-02  19:00 CON LUZ  Tenis - Cancha 2  APROBADA    H-07   
352   1687243 2020-12-02  18:00 CON LUZ  Tenis - Cancha 1  APROBADA    M-03   
353   1652644 2020-11-26       16:00:00  Tenis - Cancha 1  APROBADA    N-09   

    Unnamed: 6  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  
..         ...  
349        NaN  
350        NaN  
351        NaN  
352        NaN  
353        NaN  

[354 rows x 7 columns]

Warning: A column seems to be left over. Name inconsistency and full of "NaN". Correction needed. (Above)

<h4>Statistical Summary</h4>
df.describe()

In [7]:
df.describe(include = 'all')

/tmp/ipykernel_697/4220128335.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df.describe(include = 'all')


reserva                 fecha           hora          servicio  \
count   3.540000e+02                  354            354               354   
unique           NaN                  262             13                 2   
top              NaN  2023-03-14 00:00:00  18:00 CON LUZ  Tenis - Cancha 1   
freq             NaN                   14             63               309   
first            NaN  2020-11-26 00:00:00            NaN               NaN   
last             NaN  2023-07-08 00:00:00            NaN               NaN   
mean    5.279749e+06                  NaN            NaN               NaN   
std     2.188869e+06                  NaN            NaN               NaN   
min     3.700000e+01                  NaN            NaN               NaN   
25%     3.639766e+06                  NaN            NaN               NaN   
50%     5.117049e+06                  NaN            NaN               NaN   
75%     7.582476e+06                  NaN            NaN               NaN   
max     8.219516e+06                  NaN            NaN               NaN   

        mza-lote  casa                            Unnamed: 6  
count        354   353                                     2  
unique         1    43                                     1  
top     APROBADA  G-07  18 a 24hs. (Únicamente LOS DOMINGOS)  
freq         354    57                                     2  
first        NaN   NaN                                   NaN  
last         NaN   NaN                                   NaN  
mean         NaN   NaN                                   NaN  
std          NaN   NaN                                   NaN  
min          NaN   NaN                                   NaN  
25%          NaN   NaN                                   NaN  
50%          NaN   NaN                                   NaN  
75%          NaN   NaN                                   NaN  
max          NaN   NaN                                   NaN

<h2>Evaluating for Missing Data</h2>

The missing values are converted by default. We use the following functions to identify these missing values. There are two methods to detect missing data:
<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.


<h3 id="deal_missing_values">Deal with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>Drop data<br>
        a. Drop the whole row<br>
        b. Drop the whole column
    </li>
    <li>Replace data<br>
        a. Replace it by mean<br>
        b. Replace it by frequency<br>
        c. Replace it based on other functions
    </li>
</ol>

In [8]:
# New dataframe with rows true/false that have null elements
missing_data = df.isnull()
missing_data.head(5)
# True means is null, False the opossite

reserva   fecha   hora  servicio  mza-lote   casa  Unnamed: 6
0     False  False  False     False     False  False        True
1     False  False  False     False     False  False        True
2     False  False  False     False     False  False        True
3     False  False  False     False     False  False        True
4     False  False  False     False     False  False        True

Warning: 'Unnamed 6' full of null values.

In [9]:
# Check if there is null values in "all data frame"
df.isna().any(axis=0)

reserva       False
fecha         False
hora          False
servicio      False
mza-lote      False
casa           True
Unnamed: 6     True
dtype: bool

Warning: There is null values in casa and 'Unnamed: 6' column. (Above)

In [10]:
# Find unnamed columns
unnamed_columns = df.columns[df.columns == '']
unnamed_columns
# Si todas las columnas tienen nombre retorna false Index([], dtype='object'

Index([], dtype='object')

In [11]:
# Find named columns
unnamed_columns = df.columns[df.columns != '']
unnamed_columns
# Si todas las columnas tienen nombre retorna los nombres de las columnas, dtype='object'

Index(['reserva ', 'fecha', 'hora', 'servicio', 'mza-lote', 'casa',
       'Unnamed: 6'],
      dtype='object')

In [12]:
# Check if any column has null values and sum. Sum all null values per column.
df.isna().sum(axis=0)

reserva         0
fecha           0
hora            0
servicio        0
mza-lote        0
casa            1
Unnamed: 6    352
dtype: int64

Warning: The 'Unnamed 6' column is seen to be either empty or full of NaNs once again.

In [13]:
# Check the columns names
df.columns

Index(['reserva ', 'fecha', 'hora', 'servicio', 'mza-lote', 'casa',
       'Unnamed: 6'],
      dtype='object')

In [14]:
# Check the name of the column before delete it
print(df['Unnamed: 6'].head())

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: Unnamed: 6, dtype: object


Warning: 'Unnamed 6' is unnecessary. Proceed to delete.

In [15]:
# Drop the Empty column
df = df.drop('Unnamed: 6', axis=1)

In [16]:
# Check the columns that still exist
df.columns

Index(['reserva ', 'fecha', 'hora', 'servicio', 'mza-lote', 'casa'], dtype='object')

In [17]:
# Check if any column has null values and sum again. Sum the nulls values per column.
df.isna().sum(axis=0)

reserva     0
fecha       0
hora        0
servicio    0
mza-lote    0
casa        1
dtype: int64

Warning: 'casa' has a null value. (Above)

In [18]:
df['casa'].describe()

count      353
unique      43
top       G-07
freq        57
Name: casa, dtype: object

Warning: It is seen that there is a null element that 353 is one less than 354 (the total of the rows) (Above)

Looking for NaN element in 'casa' using boolean mask

In [19]:
# Filter obtaining null element row of the dataframe
df_row_with_null = df[df['casa'].isnull()]

In [20]:
# Show the row with null in casa
df_row_with_null

reserva       fecha           hora          servicio  mza-lote casa
306   2753105 2021-04-05  19:00 CON LUZ  Tenis - Cancha 1  APROBADA  NaN

Warning: Row 306 in 'casa' has the unique NaN value (Above)

In [21]:
# Build a Dataframe with no NaN. The new data frame without NaN is named df_without_null
df_without_null =df[df['casa'].notnull()]

In [22]:
df_without_null

reserva       fecha           hora          servicio  mza-lote    casa
0     8219516 2023-07-08       16:00:00  Tenis - Cancha 2  APROBADA    M-03
1     8218607 2023-07-08       16:00:00  Tenis - Cancha 1  APROBADA    F-04
2     8217992 2023-07-08       13:00:00  Tenis - Cancha 1  APROBADA    G-12
3     8212103 2023-07-07       11:00:00  Tenis - Cancha 1  APROBADA    A-52
4     8196830 2023-07-04       16:00:00  Tenis - Cancha 1  APROBADA    K-04
..        ...        ...            ...               ...       ...     ...
349   1712511 2020-12-04  18:00 CON LUZ  Tenis - Cancha 1  APROBADA  PRUEBA
350   1712222 2020-12-04       17:00:00  Tenis - Cancha 1  APROBADA    E-08
351   1687608 2020-12-02  19:00 CON LUZ  Tenis - Cancha 2  APROBADA    H-07
352   1687243 2020-12-02  18:00 CON LUZ  Tenis - Cancha 1  APROBADA    M-03
353   1652644 2020-11-26       16:00:00  Tenis - Cancha 1  APROBADA    N-09

[353 rows x 6 columns]

Warning: It's seen that there is 353 not null values in df_without_null. This is becouse there is only one null value in df. The sum is the total rows of the df.

In [23]:
# Check if any column has null values and sum
df_without_null.isna().sum(axis=0)

reserva     0
fecha       0
hora        0
servicio    0
mza-lote    0
casa        0
dtype: int64

In [24]:
# Set df (original) with df_without_null.
df= df_without_null

# Set df with df_without_null, becouse tpyping df is short and easy to work
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 353 entries, 0 to 353
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   reserva   353 non-null    int64         
 1   fecha     353 non-null    datetime64[ns]
 2   hora      353 non-null    object        
 3   servicio  353 non-null    object        
 4   mza-lote  353 non-null    object        
 5   casa      353 non-null    object        
dtypes: datetime64[ns](1), int64(1), object(4)
memory usage: 19.3+ KB


<b>Missing Data Dealing Successfull: NaN eliminated.</b>

# Correct data format
###### See references in: https://pandas.pydata.org/docs/user_guide/basics.html#dtypes

<p>Next step in data cleaning is checking and making sure that all data is in the correct format (int, float, text or other).</p>
​
I use:
<ol>
    <li><b>.astype()</b></li>
    <li><b>.dtype()</b></li>
</ol>

In [25]:
# Check data types
df.dtypes

reserva              int64
fecha       datetime64[ns]
hora                object
servicio            object
mza-lote            object
casa                object
dtype: object

<p>As we can see above, some columns are not of the correct data type.</p>
<p>Numerical variables should have type 'float' or 'int', and variables with strings such as categories should have type 'object'.</p>
<p> 'reserva', 'fecha' are ok.</p>
<P>I will convert data types into a proper format for each column using the "astype()" method.</p> 

#### Date and time correction is needed
Es necesario corregir fecha y hora de alquiler de las canchas.


In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 353 entries, 0 to 353
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   reserva   353 non-null    int64         
 1   fecha     353 non-null    datetime64[ns]
 2   hora      353 non-null    object        
 3   servicio  353 non-null    object        
 4   mza-lote  353 non-null    object        
 5   casa      353 non-null    object        
dtypes: datetime64[ns](1), int64(1), object(4)
memory usage: 19.3+ KB


##### Warning: Time depuration is needed. Then Join Dates and Times.
###### Fecha y hora están separadas. Fecha está en formato correcto pero sin la hora. Hora está con formato incorrecto. Hora tiene ademas la indicación de CON LUZ.
###### Es necesario juntar quitar "CON LUZ" de hora, juntar fecha y hora y dejarlas en formato datetime64[ns]

With df['fecha']:  Can see date without  hours. Se pueden ver las fechas sin las horas.

Convierto la columna de fechas en string para luego juntarla con la columna de horas.

In [27]:
# Re-Check 'hora' type to see if it is OK to join with dates or NOT OK
df['hora'].astype(str)

0           16:00:00
1           16:00:00
2           13:00:00
3           11:00:00
4           16:00:00
           ...      
349    18:00 CON LUZ
350         17:00:00
351    19:00 CON LUZ
352    18:00 CON LUZ
353         16:00:00
Name: hora, Length: 353, dtype: object

Warning: Observe "CON LUZ". Extraction and spliting is needed. (Above)

In [28]:
split_hora = df['hora'].astype(str)

In [29]:
splited_hora = split_hora.str.split(" ", expand= True)

In [30]:
splited_hora

0     1     2
0    16:00:00  None  None
1    16:00:00  None  None
2    13:00:00  None  None
3    11:00:00  None  None
4    16:00:00  None  None
..        ...   ...   ...
349     18:00   CON   LUZ
350  17:00:00  None  None
351     19:00   CON   LUZ
352     18:00   CON   LUZ
353  16:00:00  None  None

[353 rows x 3 columns]

In [31]:
# Select Times in a shora serie
shora = splited_hora[0].astype("string")

In [32]:
# Check if shora conteins time series
shora

0      16:00:00
1      16:00:00
2      13:00:00
3      11:00:00
4      16:00:00
         ...   
349       18:00
350    17:00:00
351       19:00
352       18:00
353    16:00:00
Name: 0, Length: 353, dtype: string

In [33]:
# Convert 'fecha' in str format
sfecha = df['fecha'].astype("string")

In [34]:
# Check if sfecha is a correct serie of dates
sfecha

0      2023-07-08
1      2023-07-08
2      2023-07-08
3      2023-07-07
4      2023-07-04
          ...    
349    2020-12-04
350    2020-12-04
351    2020-12-02
352    2020-12-02
353    2020-11-26
Name: fecha, Length: 353, dtype: string

In [35]:
# Concat the sfecha with shora
s_fechayhora = pd.DataFrame(sfecha + " " + shora)

In [36]:
# Check if concatenation is well done
s_fechayhora

0
0    2023-07-08 16:00:00
1    2023-07-08 16:00:00
2    2023-07-08 13:00:00
3    2023-07-07 11:00:00
4    2023-07-04 16:00:00
..                   ...
349     2020-12-04 18:00
350  2020-12-04 17:00:00
351     2020-12-02 19:00
352     2020-12-02 18:00
353  2020-11-26 16:00:00

[353 rows x 1 columns]

In [37]:
# Convert s_fechayhora to datetime60[ns] serie
fechayhora = pd.to_datetime(s_fechayhora[0])

In [38]:
# Check if serie of date and times is ok
fechayhora

0     2023-07-08 16:00:00
1     2023-07-08 16:00:00
2     2023-07-08 13:00:00
3     2023-07-07 11:00:00
4     2023-07-04 16:00:00
              ...        
349   2020-12-04 18:00:00
350   2020-12-04 17:00:00
351   2020-12-02 19:00:00
352   2020-12-02 18:00:00
353   2020-11-26 16:00:00
Name: 0, Length: 353, dtype: datetime64[ns]

Date and Time Ready. Time to update the Dataframe under Analysis.

Fecha y Hora listas para actualizar en el Dataframe que se está analizando.

In [39]:
# Set the date and time serie into a data frame
df['fecha'] = fechayhora

In [40]:
# Check if column fecha is updated with date and time (not only date)
df['fecha']

0     2023-07-08 16:00:00
1     2023-07-08 16:00:00
2     2023-07-08 13:00:00
3     2023-07-07 11:00:00
4     2023-07-04 16:00:00
              ...        
349   2020-12-04 18:00:00
350   2020-12-04 17:00:00
351   2020-12-02 19:00:00
352   2020-12-02 18:00:00
353   2020-11-26 16:00:00
Name: fecha, Length: 353, dtype: datetime64[ns]

In [41]:
# Rename the 'fecha' column with 'fechayhora' new name
#        Warning!! df['fecha'] = df['fecha'].rename('fechayhora', inplace = True) does not work.
#        Force renaming with df.rename method and inplace = True)
df.rename(columns = {'fecha' : 'fechayhora'}, inplace = True)

Warning: There is no need of hora anymore. fechayhora contains date and times.

In [42]:
# Delete 'hora' column
df = df.drop(columns=['hora'])

In [43]:
df.head()

reserva           fechayhora          servicio  mza-lote  casa
0   8219516 2023-07-08 16:00:00  Tenis - Cancha 2  APROBADA  M-03
1   8218607 2023-07-08 16:00:00  Tenis - Cancha 1  APROBADA  F-04
2   8217992 2023-07-08 13:00:00  Tenis - Cancha 1  APROBADA  G-12
3   8212103 2023-07-07 11:00:00  Tenis - Cancha 1  APROBADA  A-52
4   8196830 2023-07-04 16:00:00  Tenis - Cancha 1  APROBADA  K-04

<b>Date and Time Correction Successfull!</b> 
<b> Next "CON LUZ" binning process.</b> 

#### New column 'ilumina' setup

In [44]:
# Build new column from splited_hora
splited_hora

0     1     2
0    16:00:00  None  None
1    16:00:00  None  None
2    13:00:00  None  None
3    11:00:00  None  None
4    16:00:00  None  None
..        ...   ...   ...
349     18:00   CON   LUZ
350  17:00:00  None  None
351     19:00   CON   LUZ
352     18:00   CON   LUZ
353  16:00:00  None  None

[353 rows x 3 columns]

In [45]:
# Take a column of splited_hora and complete with 'SOL' where is empty/None
sluz = splited_hora[2].astype("string")
sluz.fillna("SOL", inplace = True)

# Make a categorical convertion from sluz to cluz
cluz = pd.Categorical(sluz, categories=["SOL", "LUZ"], ordered=False)

# Set the new column in df with iluminacion from cluz
df['iluminacion'] = cluz

In [46]:
# Check the resulting dataframe
df.head()

reserva           fechayhora          servicio  mza-lote  casa iluminacion
0   8219516 2023-07-08 16:00:00  Tenis - Cancha 2  APROBADA  M-03         SOL
1   8218607 2023-07-08 16:00:00  Tenis - Cancha 1  APROBADA  F-04         SOL
2   8217992 2023-07-08 13:00:00  Tenis - Cancha 1  APROBADA  G-12         SOL
3   8212103 2023-07-07 11:00:00  Tenis - Cancha 1  APROBADA  A-52         SOL
4   8196830 2023-07-04 16:00:00  Tenis - Cancha 1  APROBADA  K-04         SOL

<b>Ilumina CON/SIN LUZ correction Successfull!</b> 
<b>Next Cancha/Servicio depuration</b> 

#### New column Cancha Setup

In [47]:
# Servicio has merged Tenis and Cancha with number. Split and correction is necesary
# Split column servicio to take only the number of the cancha
splited_servicio = df['servicio'].str.split(" ", expand= True)

In [48]:
splited_servicio

0  1       2  3
0    Tenis  -  Cancha  2
1    Tenis  -  Cancha  1
2    Tenis  -  Cancha  1
3    Tenis  -  Cancha  1
4    Tenis  -  Cancha  1
..     ... ..     ... ..
349  Tenis  -  Cancha  1
350  Tenis  -  Cancha  1
351  Tenis  -  Cancha  2
352  Tenis  -  Cancha  1
353  Tenis  -  Cancha  1

[353 rows x 4 columns]

In [49]:
# Reunite Cancha with number of the cancha in a serie scancha
scancha = splited_servicio[2] + " " + splited_servicio [3]

In [50]:
# Asign to servicio categorical serie of scancha
df['servicio'] = pd.Categorical(scancha, categories=["Cancha 1", "Cancha 2"], ordered=False)

In [51]:
df.head()

reserva           fechayhora  servicio  mza-lote  casa iluminacion
0   8219516 2023-07-08 16:00:00  Cancha 2  APROBADA  M-03         SOL
1   8218607 2023-07-08 16:00:00  Cancha 1  APROBADA  F-04         SOL
2   8217992 2023-07-08 13:00:00  Cancha 1  APROBADA  G-12         SOL
3   8212103 2023-07-07 11:00:00  Cancha 1  APROBADA  A-52         SOL
4   8196830 2023-07-04 16:00:00  Cancha 1  APROBADA  K-04         SOL

<b>Cancha 1/Cancha2 servicio correction Successfull!</b> 
<b>Next Manzana/Lote depuration</b> 

#### Mza-lote Depuration - Categorization

In [52]:
# Categorization of the aprobation
df['mza-lote'] = pd.Categorical(df['mza-lote'], categories=["APROBADA", "NO APROBADA"], ordered=False)

In [53]:
df.dtypes

reserva                 int64
fechayhora     datetime64[ns]
servicio             category
mza-lote             category
casa                   object
iluminacion          category
dtype: object

#### Columns organization

In [54]:
# Reordering coluns: Inster ilumina column to replace inuminacion column between fechayhora and servicio
# df.reindex(columns = ['reserva','fechayhora','iluminacion','mza-lote','casa']) # Ojo: esto no anda bien reserva queda toda en NaN!
df.insert(2, 'ilumina', df['iluminacion'], True)

# Reordering columns: Drop the Iluminacion column
df = df.drop('iluminacion', axis=1)

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 353 entries, 0 to 353
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   reserva     353 non-null    int64         
 1   fechayhora  353 non-null    datetime64[ns]
 2   ilumina     353 non-null    category      
 3   servicio    353 non-null    category      
 4   mza-lote    353 non-null    category      
 5   casa        353 non-null    object        
dtypes: category(3), datetime64[ns](1), int64(1), object(1)
memory usage: 12.4+ KB


<h4>Reorder df by date and time</h4>

In [56]:
df.sort_values(by = 'fechayhora', inplace = True)

In [57]:
df

reserva           fechayhora ilumina  servicio  mza-lote    casa
353   1652644 2020-11-26 16:00:00     SOL  Cancha 1  APROBADA    N-09
352   1687243 2020-12-02 18:00:00     LUZ  Cancha 1  APROBADA    M-03
351   1687608 2020-12-02 19:00:00     LUZ  Cancha 2  APROBADA    H-07
350   1712222 2020-12-04 17:00:00     SOL  Cancha 1  APROBADA    E-08
349   1712511 2020-12-04 18:00:00     LUZ  Cancha 1  APROBADA  PRUEBA
..        ...                 ...     ...       ...       ...     ...
4     8196830 2023-07-04 16:00:00     SOL  Cancha 1  APROBADA    K-04
3     8212103 2023-07-07 11:00:00     SOL  Cancha 1  APROBADA    A-52
2     8217992 2023-07-08 13:00:00     SOL  Cancha 1  APROBADA    G-12
1     8218607 2023-07-08 16:00:00     SOL  Cancha 1  APROBADA    F-04
0     8219516 2023-07-08 16:00:00     SOL  Cancha 2  APROBADA    M-03

[353 rows x 6 columns]

#### Adding users column

In [58]:
# Generate a random Series with 354 elements, where each element is either "Propietario" or "Invitado"
choices = ["Propietario", "Invitado"]
random_data = np.random.choice(choices, size=354, replace=True)

# Create a pandas Series from the random data
random_series = pd.Series(random_data)

# Create a column with de serie
df['usuario'] = random_series

In [59]:
df

reserva           fechayhora ilumina  servicio  mza-lote    casa  \
353   1652644 2020-11-26 16:00:00     SOL  Cancha 1  APROBADA    N-09   
352   1687243 2020-12-02 18:00:00     LUZ  Cancha 1  APROBADA    M-03   
351   1687608 2020-12-02 19:00:00     LUZ  Cancha 2  APROBADA    H-07   
350   1712222 2020-12-04 17:00:00     SOL  Cancha 1  APROBADA    E-08   
349   1712511 2020-12-04 18:00:00     LUZ  Cancha 1  APROBADA  PRUEBA   
..        ...                 ...     ...       ...       ...     ...   
4     8196830 2023-07-04 16:00:00     SOL  Cancha 1  APROBADA    K-04   
3     8212103 2023-07-07 11:00:00     SOL  Cancha 1  APROBADA    A-52   
2     8217992 2023-07-08 13:00:00     SOL  Cancha 1  APROBADA    G-12   
1     8218607 2023-07-08 16:00:00     SOL  Cancha 1  APROBADA    F-04   
0     8219516 2023-07-08 16:00:00     SOL  Cancha 2  APROBADA    M-03   

         usuario  
353     Invitado  
352  Propietario  
351     Invitado  
350  Propietario  
349     Invitado  
..           ...  
4    Propietario  
3    Propietario  
2       Invitado  
1    Propietario  
0    Propietario  

[353 rows x 7 columns]

In [60]:
df['usuario'] = pd.Categorical(df['usuario'], categories=["Propietario", "Invitado"], ordered=False)

## Append new column for cancha's princing: Create price billed

In [61]:
df['precio'] = 0.0

In [62]:
df.dtypes

reserva                int64
fechayhora    datetime64[ns]
ilumina             category
servicio            category
mza-lote            category
casa                  object
usuario             category
precio               float64
dtype: object

### Pricing Table

In [63]:
# Lee el archivo excel con pandas y guardar el resultado en un dataframe
dfp = pd.read_excel("PreciosCanchasTenis2.0.xlsx")

In [64]:
dfp

Fecha      Usuario Ilumina  Precio
0  2021-01-01  Propietario     SOL     100
1  2021-01-01  Propietario     LUZ     200
2  2021-01-01     Invitado     SOL     100
3  2021-01-01     Invitado     LUZ     200
4  2023-02-01  Propietario     SOL     150
5  2023-02-01  Propietario     LUZ     250
6  2023-02-01     Invitado     SOL     250
7  2023-02-01     Invitado     LUZ     400
8  2023-07-01  Propietario     SOL     200
9  2023-07-01  Propietario     LUZ     300
10 2023-07-01     Invitado     SOL     300
11 2023-07-01     Invitado     LUZ     500

In [65]:
dfp.dtypes

Fecha      datetime64[ns]
Usuario            object
Ilumina            object
Precio              int64
dtype: object

In [66]:
dfp['Precio'] = dfp['Precio'].astype(float)
dfp['Usuario'] = pd.Categorical(dfp['Usuario'], categories=["Propietario", "Invitado"], ordered=False)
dfp['Ilumina'] = pd.Categorical(dfp['Ilumina'], categories=["SOL","LUZ"], ordered=False)

In [67]:
dfp.dtypes

Fecha      datetime64[ns]
Usuario          category
Ilumina          category
Precio            float64
dtype: object

In [68]:
dfp

Fecha      Usuario Ilumina  Precio
0  2021-01-01  Propietario     SOL   100.0
1  2021-01-01  Propietario     LUZ   200.0
2  2021-01-01     Invitado     SOL   100.0
3  2021-01-01     Invitado     LUZ   200.0
4  2023-02-01  Propietario     SOL   150.0
5  2023-02-01  Propietario     LUZ   250.0
6  2023-02-01     Invitado     SOL   250.0
7  2023-02-01     Invitado     LUZ   400.0
8  2023-07-01  Propietario     SOL   200.0
9  2023-07-01  Propietario     LUZ   300.0
10 2023-07-01     Invitado     SOL   300.0
11 2023-07-01     Invitado     LUZ   500.0

### DataFrames Inconcistency - Dates Rows

The use of courtyards starts on year 2020 while pricing references information starts on year 2020.
El uso de las canchas comienza en el año 2020 mientras que la información de los precios inicia en el año 2021.

Insert a new rows to start pricing from 2020-01-01.
Insertar nuevas filas para iniciar con los precios 01-01-2020.

In [69]:
# Insert 4 rows - Prepare new 4 rows to add
data = {
    'Fecha': ['2020-01-01', '2020-01-01', '2020-01-01', '2020-01-01'],
    'Usuario': ['Propietario', 'Propietario', 'Invitado', 'Invitado'],
    'Ilumina': ['SOL', 'LUZ', 'SOL', 'LUZ'],
    'Precio': [90, 190, 90, 190]
}

dfp2add = pd.DataFrame(data)

In [70]:
# it's necesary cast the columns
dfp2add['Fecha']= pd.to_datetime(dfp2add['Fecha'])
dfp2add['Usuario']= dfp2add['Usuario'].astype('category')
dfp2add['Ilumina']= dfp2add['Ilumina'].astype('category')
dfp2add['Precio']=dfp2add['Precio'].astype(float)

In [71]:
dfp2add

Fecha      Usuario Ilumina  Precio
0 2020-01-01  Propietario     SOL    90.0
1 2020-01-01  Propietario     LUZ   190.0
2 2020-01-01     Invitado     SOL    90.0
3 2020-01-01     Invitado     LUZ   190.0

In [72]:
# Insert 4 rows - Add new rows at the begining
frames = [dfp2add, dfp]      # build a list of dataframes to concat in order
result = pd.concat(frames)

In [73]:
result.dtypes

Fecha      datetime64[ns]
Usuario          category
Ilumina          category
Precio            float64
dtype: object

In [74]:
dfp = result

### Canchas Inconcistency - User Category Not Found

#### El archivo de canchas no especifica si los usuarios son propietarios o invitados

In [75]:
# Generate a random Series with 354 elements, where each element is either "Propietario" or "Invitado"
choices = ["Propietario", "Invitado"]
random_data = np.random.choice(choices, size=354, replace=True)

# Create a pandas Series from the random data
random_series = pd.Series(random_data)

In [76]:
random_series

0         Invitado
1      Propietario
2      Propietario
3      Propietario
4         Invitado
          ...     
349    Propietario
350       Invitado
351       Invitado
352       Invitado
353       Invitado
Length: 354, dtype: object

In [77]:
df['usuario'] = random_series

In [78]:
df

reserva           fechayhora ilumina  servicio  mza-lote    casa  \
353   1652644 2020-11-26 16:00:00     SOL  Cancha 1  APROBADA    N-09   
352   1687243 2020-12-02 18:00:00     LUZ  Cancha 1  APROBADA    M-03   
351   1687608 2020-12-02 19:00:00     LUZ  Cancha 2  APROBADA    H-07   
350   1712222 2020-12-04 17:00:00     SOL  Cancha 1  APROBADA    E-08   
349   1712511 2020-12-04 18:00:00     LUZ  Cancha 1  APROBADA  PRUEBA   
..        ...                 ...     ...       ...       ...     ...   
4     8196830 2023-07-04 16:00:00     SOL  Cancha 1  APROBADA    K-04   
3     8212103 2023-07-07 11:00:00     SOL  Cancha 1  APROBADA    A-52   
2     8217992 2023-07-08 13:00:00     SOL  Cancha 1  APROBADA    G-12   
1     8218607 2023-07-08 16:00:00     SOL  Cancha 1  APROBADA    F-04   
0     8219516 2023-07-08 16:00:00     SOL  Cancha 2  APROBADA    M-03   

         usuario  precio  
353     Invitado     0.0  
352     Invitado     0.0  
351     Invitado     0.0  
350     Invitado     0.0  
349  Propietario     0.0  
..           ...     ...  
4       Invitado     0.0  
3    Propietario     0.0  
2    Propietario     0.0  
1    Propietario     0.0  
0       Invitado     0.0  

[353 rows x 8 columns]

In [79]:
# Convert dates column in df and dfp to datetime if it's not already in datetime format
df['fechayhora'] = pd.to_datetime(df['fechayhora'])
dfp['Fecha'] = pd.to_datetime(dfp['Fecha'])

# Sort dfp by date to ensure consecutive dates are next to each other
dfp = dfp.sort_values(by='Fecha')

# Create a Dataframe with uniques start and end datas of dfp
serie_unique = dfp['Fecha'].unique()                       # gest the array of unique dates
dfp_u = pd.DataFrame({
        'FecIni' : serie_unique, 'FecFin': serie_unique    # initially both dfp_u columns has the samedates
})

# Configure end dates with shift and append on the las row with last rown from the unique date.
dfp_u['FecFin'] = dfp_u['FecFin'].shift(-1)   # shit values up             
n = len(dfp_u)-1
dfp_u['FecFin'][n] = pd.Timestamp.today()  # complete the last value


# Iterate over rows in df DataFrame
i=0
for df_row in df.itertuples():
    
    # Get the current date, user, and illuminate from the row
    date = getattr(df_row,'fechayhora')
    user = getattr(df_row, 'usuario')
    illuminate = getattr(df_row,'ilumina')
    i=i+1
    # debug: print("df fila:", getattr(row, 'Index'))
    
    # Initialize the precio to 0.0 for the current row
    precio = 0.0
    
    # Iterate over consecutive pairs of dates in dfp
    for dfp_row in dfp_u.itertuples():
        start_date = getattr(dfp_row, 'FecIni')
        end_date = getattr(dfp_row, 'FecFin')
        
        # Check if the current date falls between the start and end dates
        if start_date <= date <= end_date :
            # debug: print("dfp index =", getattr(row, 'Index'), "Si! ", start_date, " <= ", date, " < ",  end_date)
            
            # Filter dfp based on the conditions for merging
            dfp_filtered = dfp[
                (dfp['Fecha'] == start_date) &
                (dfp['Usuario'] == user) &
                (dfp['Ilumina'] == illuminate)
            ]

            # Set the 'precio' value in df based on the filtered dfp value
            if not dfp_filtered.empty:
                df.at[getattr(df_row, 'Index'), 'precio'] = dfp_filtered.iloc[0]['Precio']
            else:
                print("no hay precio", df_row, Index)

In [80]:
df

reserva           fechayhora ilumina  servicio  mza-lote    casa  \
353   1652644 2020-11-26 16:00:00     SOL  Cancha 1  APROBADA    N-09   
352   1687243 2020-12-02 18:00:00     LUZ  Cancha 1  APROBADA    M-03   
351   1687608 2020-12-02 19:00:00     LUZ  Cancha 2  APROBADA    H-07   
350   1712222 2020-12-04 17:00:00     SOL  Cancha 1  APROBADA    E-08   
349   1712511 2020-12-04 18:00:00     LUZ  Cancha 1  APROBADA  PRUEBA   
..        ...                 ...     ...       ...       ...     ...   
4     8196830 2023-07-04 16:00:00     SOL  Cancha 1  APROBADA    K-04   
3     8212103 2023-07-07 11:00:00     SOL  Cancha 1  APROBADA    A-52   
2     8217992 2023-07-08 13:00:00     SOL  Cancha 1  APROBADA    G-12   
1     8218607 2023-07-08 16:00:00     SOL  Cancha 1  APROBADA    F-04   
0     8219516 2023-07-08 16:00:00     SOL  Cancha 2  APROBADA    M-03   

         usuario  precio  
353     Invitado    90.0  
352     Invitado   190.0  
351     Invitado   190.0  
350     Invitado    90.0  
349  Propietario   190.0  
..           ...     ...  
4       Invitado   300.0  
3    Propietario   200.0  
2    Propietario   200.0  
1    Propietario   200.0  
0       Invitado   300.0  

[353 rows x 8 columns]

## Save and Close File depurated

In [82]:
fname_out = "Canchas 0.2 Depurado"

In [83]:
df.to_excel(fname_out, index_label="label", merge_cells=False)

ValueError: No engine for filetype: '2 Depurado'